use-location.js

```diff
@@ -9,55 +13,46 @@ const eventPushState = "pushState";
-  const [{ path }, update] = useState(() => ({ path: currentPathname(base) }));
-  // @see https://reactjs.org/docs/hooks-reference.html#lazy-initial-state
-  const prevHash = useRef(path + location.search);
-
-  useEffect(() => {
-    // this function checks if the location has been changed since the
-    // last render and updates the state only when needed.
-    // unfortunately, we can't rely on `path` value here, since it can be stale,
-    // that's why we store the last pathname in a ref.
-    const checkForUpdates = () => {
-      const pathname = currentPathname(base);
-      const hash = pathname + location.search;
-
-      if (prevHash.current !== hash) {
-        prevHash.current = hash;
-        update({ path: pathname });
-      }
-    };
-
-    events.forEach((e) => addEventListener(e, checkForUpdates));
+export const subscribeToLocation = (callback) => {
+  for (const event of events) {
+    window.addEventListener(event, callback);
+  }
+  return () => {
+    for (const event of events) {
+      window.removeEventListener(event, callback);
+    }
+  };
+};
 
-    // it's possible that an update has occurred between render and the effect handler,
-    // so we run additional check on mount to catch these updates. Based on:
-    // https://gist.github.com/bvaughn/e25397f70e8c65b0ae0d7c90b731b189
-    checkForUpdates();
+const currentSearch = () => location.search;
+export const useSearch = () =>
+  useSyncExternalStore(subscribeToLocation, currentSearch);
 
-    return () => events.forEach((e) => removeEventListener(e, checkForUpdates));
-  }, [base]);
+export const usePathname = (base = "") =>
+  useSyncExternalStore(subscribeToLocation, () => currentPathname(base));
 
-  // the 2nd argument of the `useLocation` return value is a function
-  // that allows to perform a navigation.
-  //
-  // the function reference should stay the same between re-renders, so that
-  // it can be passed down as an element prop without any performance concerns.
-  const navigate = useCallback(
-    (to, { replace = false } = {}) =>
-      history[replace ? eventReplaceState : eventPushState](
-        null,
-        "",
-        // handle nested routers and absolute paths
-        to[0] === "~" ? to.slice(1) : base + to
-      ),
-    [base]
+export const navigate = (to, { replace = false } = {}, base = "") =>
+  history[replace ? eventReplaceState : eventPushState](
+    null,
+    "",
+    to[0] === "~" ? to.slice(1) : base + to
   );
 
-  return [path, navigate];
+// the 2nd argument of the `useLocation` return value is a function
+// that allows to perform a navigation.
+//
+// the function reference should stay the same between re-renders, so that
+// it can be passed down as an element prop without any performance concerns.
+export const useNavigate = (base = "") => {
+  const [nav] = useState([base, (to, opts) => navigate(to, opts, nav[0])]);
+  useIsomorphicLayoutEffect(() => {
+    nav[0] = base;
+  });
+  return nav[1];
```

Comment:
Could we rewrite it to just:

```js
export const useNavigate = (base = "") => 
  useCallback((to, opts) => navigate(to, opts, base), [base])
```

There must be some reason to use your version, but I just can figure out what it is 😄

Reply:
@molefrog yes, in fact there 2 reasons: 

1. My version won't unmemoize every time `base` changes, causing unnecessary re-rendering. The function returned from this hook is guaranteed to remain constant for the entire lifespan of this hook. (And in fact, I don't believe that even `useCallback` with an empty dependency array guarantees this since I believe it is implemented in terms of `useMemo`).

2. Since we are passing base in from a router, which is effectively a _ref_ (value that is mutated without triggering a re-render), that means that we are depending on the value of a ref in the callback dependency array, which you are not supposed to do, or at least it is nowhere documented that you can do this safely so it makes me very uneasy. See, for example: https://stackoverflow.com/questions/60476155/is-it-safe-to-use-ref-current-as-useeffects-dependency-when-ref-points-to-a-dom

Reply:
Ooh, but in talking about this, I just realized that I do need to make one small change here 😅 😅 😅

Reply:
Regarding 2), maybe I'm missing something afaik React compares dependencies using `===` and since `base` is supposed to be always a string the callback will be up to date.

Reply:
@molefrog true perhaps, but the navigate function itself won't be up to date... since changing base does not trigger a re-render, it's possible that this callback won't ever be updated, leading to a situation where clicking a button, for example, sends you to the wrong path since the navigate function refers to `base` _as it was when the callback was memoized_, but not as it is now. That was actually the thing I fixed after saying "I do need to make one small change here".

Reply:
And to be clear: when I say "changing base does not trigger a re-render", I'm referring to the re-render of child components within that component tree that are potentially using this navigate hook. Anything that is Memo'd won't re-render as a result of the re-render of the ancestor due to base changing in the first place. Hopefully that makes sense.

Reply:
Now, one option that I considered is similar to what you suggest:

```typescript
export const useNavigate = (opts) => 
  useCallback((to, o) => navigate(to, o, opts.base || ""), [opts]);
```

That would work just fine. However it would fail for cases where the user is constructing `opts` dynamically, e.g.:
`useNavigate({base})` (fails) ❌   
`useNavigate(useRouter())` (works) ✅ 

Which is the reverse of the current logic in this library:
`useNavigate({base})` (works--provided `base` is a prop, not a ref) ✅ 
`useNavigate(useRouter())` (fails) ❌   

And thus might present backwards-incompatibility issues.

So instead of that, I opted for my current strategy which is the only one that works for _both_ usage patterns.

Reply:
@molefrog any thoughts on the above?

Reply:
Fair point, think we can resolve this! 👍